### PR TITLE
Fix dependency on batch size by using background instead of minimum

### PR DESCRIPTION
**Pull request by dereklhansen**
Changes in this PR:
- Within LocationEncoder and BinaryEncoder, instead of subtracting `image_ptiles.min()`, we subtract a background value that is given as input. 
- For simulated data, the simulated dataset provides the true background values
- The decoder now only returns the rendered image without background and noise added. We also remove all background-related methods from decoder.
- The decoder does not return a separate variance image
- The SDSS blended galaxies dataset uses background values provided by SDSS
- inference.py requires background as an input (which is provided by SDSS)

### CODE REVIEW

**Comment by dereklhansen** [comment]
We don't quite get as good performance on M2 when we use `background - 5 * sqrt(background)` as the minimum per-channel
```
Sleep phase TPR:  tensor(0.4623, device='cuda:0')
Sleep phase PPV:  tensor(0.3458, device='cuda:0')
```

**Comment by dereklhansen** [comment]
Part of the performance issue could be that the hubble data used to test the procedure probably has a different background than the training data.

Hubble:
```
>>> z.min(dim=1)[0].min(dim=1)[0]
tensor([ 740.6288, 1240.9622])
```
Training:
https://github.com/prob-ml/bliss/blob/5a91bf362d754366fad37625958786c7051bf699/tests/m2/m2.yaml#L32-L43

**Comment by dereklhansen** [comment]
Might want to merge this with #428

**Comment by dereklhansen** [comment]
Closes #392

**Comment by dereklhansen** [comment]
Closes #438

**Comment by dereklhansen** [comment]
PR is ready for review @ismael-mendoza if you want to take a look. I'll handle the final merge since we need to retrain the LocationEncoder and BinaryEncoder.

**Comment by ismael-mendoza** [comment]
Thanks @dereklhansen I'm taking a look now

**Review by ismael-mendoza — APPROVED** [review]
Thanks Derek! This looks good to me. I only couldn't follow the changes in `ChunkedScene` really I still need to understand that class. I just have a few questions but I don't think any really should prevent you from merging. 

Did you run it in the GPU ?

File: bliss/datasets/sdss_blended_galaxies.py [hunk 1]
```diff
@@ -136,34 +141,31 @@ def _collate(tile_catalogs: List[Dict[str, Tensor]]):
...
+            for chunk, bg in tqdm(zip(chunks, bg_chunks)):
+                chunk_device = chunk.to(self.prerender_device).unsqueeze(0)
+                bg_device = bg.to(self.prerender_device).unsqueeze(0)
+                tile_map = encoder.max_a_post(chunk_device, bg_device)
                 if tile_map["galaxy_bools"].sum() > 0:
                     catalogs.append(cpu(tile_map))
                     chunks_with_galaxies.append(chunk.cpu())
+                    bgs_with_galaxies.append(bg.cpu())
         chunks_with_galaxies = torch.stack(chunks_with_galaxies, dim=0)
+        bgs_with_galaxies = torch.stack(bgs_with_galaxies, dim=0)
```
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/datasets/sdss_blended_galaxies.py:161]: The background values can vary across the image, so we have to cut out the background at the appropriate chunk

File: bliss/datasets/simulated.py [hunk 2]
```diff
@@ -48,15 +51,17 @@ def batch_generator(self):
...
             batch = self.image_prior.sample_prior(batch_size=self.batch_size)
-            images, _ = self.image_decoder.render_images(
+            images = self.image_decoder.render_images(
                 batch["n_sources"],
                 batch["locs"],
                 batch["galaxy_bools"],
                 batch["galaxy_params"],
                 batch["fluxes"],
-                add_noise=True,
+                add_noise=False,
```
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/datasets/simulated.py:60]: Yes this option should be removed

File: bliss/encoder.py [hunk 3]
```diff
@@ -87,12 +90,15 @@ def max_a_post(self, image_ptiles: Tensor) -> Dict[str, Tensor]:
             - 'galaxy_bools', 'star_bools', and 'galaxy_probs' from BinaryEncoder.
             - 'galaxy_params' from GalaxyEncoder.
         """
-        var_params = self.location_encoder.encode(image_ptiles)
+        log_image = subtract_bg_and_log_transform(image, background)
```
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/encoder.py:93]: The transforms used for each model are the same as before. The only difference is that the LocationEncoder and BinaryEncoder now use the background instead of "image_ptiles.min()". 

I found that removing the log transform on the LocationEncoder led to a significant drop in performance on the M2 GPU test. I haven't run a comparison of different transformations on the VAE.

File: bliss/models/decoder.py [hunk 4]
```diff
@@ -132,53 +125,18 @@ def render_images(
             A tuple of the **full** image in shape (batch_size x n_bands x slen x slen) and
             its variance (same size).
         """
+        assert not add_noise
```
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/models/decoder.py:128]: Yes, this option should be removed

File: bliss/predict.py [hunk 5]
```diff
@@ -15,6 +15,7 @@
     get_images_in_tiles,
     get_is_on_from_n_sources,
     get_n_tiles_hw,
+    subtract_bg_and_log_transform,
 )
```
[the remark this answers is not in the document]
Comment (dereklhansen) [reply bliss/predict.py:20]: Yes, I just wanted to pass the tests...

**Comment by dereklhansen** [comment]
> Did you run it in the GPU ?

I checked that the M2 GPU test still trains to the threshold of accuracy. It was a little tricky since the background for the Hubble Data is higher than what we train on, so I had to estimate background for the Hubble Data by plugging in the minimum at test time.

**Comment by dereklhansen** [comment]
> I only couldn't follow the changes in `ChunkedScene` really I still need to understand that class.

The `ChunkedScene` class is still a bit WIP, but these changes should be mostly orthogonal. In fact, they make the `ChunkedScene` a little less complicated since we no longer have to subtract the background and add it back in to do the reconstruction.
